Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+  void initialize(f_type f) {
+    count_f(f);
+    allocate();
+    set_f(f, false);
+  }
+
+  /*! \brief Refresh the data according to the 'recursive function'
+   *
+   * The NdFlattener must have been constructed with a 'recursive function'.
+   * Now refresh the values with another 'recursive function'. The subarray lengths 
+   * resulting from both 'recursive functions' must coincide, as the indices arrays
+   * are not changed.
+   * 
+   * \param f - the 'recursive function'
+   * \tparam f_type - to allow for any type of the 'recursive function'
+   */
+  template<class f_type>
+  void refresh(f_type f){
+    reset_iNode();
+    set_f(f, true);
+  }
+
+protected:
+  /*! \brief Determine the space required for reading the 'recursive function'.
+   *
+   * \param f - the 'recursive function'
+   */
+  template<class f_type>
+  void count_f(f_type f) {
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      nNodes++;
+      Base::count_f(f.second(iChild));
+    }
+  }  
+
+  /*! \brief Read the 'recursive function' into the allocated arrays.
+   *
+   * \param f - the 'recursive function'
+   * \param refresh - if true, the object is already initialized and only the data
+   *   in layer 1 have to be overwritten
+   * \tparam f_type - to allow for any type of the 'recursive function'
+   */
+  template<class f_type>
+  void set_f(f_type f, bool refresh) {
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      Base::set_f(f.second(iChild), refresh);
+      if(!refresh){
+        indices[iNode+1] = indices[iNode] + f.second(iChild).first;
+      } else {
+        if( indices[iNode+1] != indices[iNode] + f.second(iChild).first ){
+          SU2_MPI::Error("NdFlattener: Structure has changed, cannot refresh.", CURRENT_FUNCTION);
+        }
+      }
+      iNode++;
+    }
+  }
+
+
+  /*=== Construct with Allgatherv ===*/
+
+public:
+  /*! \brief Initialize a flattener with K indices by combining distributed flatteners with (K-1) indices each.
+   *
+   * The new first index will encode the rank of the process. Data is exchanged in MPI::Allgatherv-style
+   * collective communication.
+   * \param[in] mpi_env - The MPI environment used for communication.
+   * \param[in] local_version - The local NdFlattener structure with (K-1) indices.
+   */
+  template<typename MPI_Environment_type>
+  void initialize( 
+    MPI_Environment_type const& mpi_env,
+    Base const* local_version 
+  ) {
+    Index** Nodes_all = new Index*[K]; // [k][r] is number of all nodes in layer (k+1), rank r in the new structure
+    for(size_t k=0; k<K; k++)
+      Nodes_all[k] = nullptr;
+    Nodes_all[K-1] = new Index[mpi_env.size]; // {1, 1, ..., 1}
+    int* displs = new int[mpi_env.size]; // {0, 1, ..., size-1}
+    int* ones = new int[mpi_env.size]; // {1,1, ...}
+    for(int r=0; r<mpi_env.size; r++){
+      nNodes += Nodes_all[K-1][r] = 1;
+      displs[r] = r;
+      ones[r] = 1;
```

Comment:
Please use vector for these as well, it makes it so much cleaner to initialize.

Nodes_all is rectangular... please use a matrix type

Reply:
done